Share build info with Provisioner and Post-Processor via HCL2 variables (#9444)
hashicorp/packer: website/pages/docs/from-1.5/special-variables.mdx

---
layout: docs

> page_title: Special Variables - HCL Configuration Language

Reviewer: I'd name these `builder contextual variables`

sidebar_title: Special Variables
description: |-
  Special variables provide connection information and basic instance state information.
  This page covers all existing special variables.
---

# Build Variables

`@include 'from-1.5/beta-hcl2-note.mdx'`

> Build variables will allow you to access, from provisioners and post-processors, connection information and basic instance state information.

Reviewer: ```suggestion
Build variables will allow you to access connection information and basic instance state information for a builder. 
-> Note: build variables are currently only available for provisioners and post-processors.
```

> All special build variables are prefixed with `build`:

Reviewer: ```suggestion
All special build variables are stored in the `build` variable:
```

```hcl
  provisioner "shell" {
      environment_vars = ["TESTVAR=${build.PackerRunUUID}"]
      inline = ["echo $TESTVAR"]
  }
```

> The valid build variables to request are:

Reviewer: ```suggestion
Here is the list of available build variables :
```

- **ID**: Represents the vm being provisioned. For example, in Amazon it is the instance id; in digitalocean,
  it is the droplet id; in Vmware, it is the vm name.

- **Host**, **Port**, **User** and **Password**: The host, port, user, and password that Packer uses to access the machine.
  Useful for using the shell local provisioner to run Ansible or Inspec against the provisioned instance.

- **ConnType**: Type of communicator being used. For example, for SSH communicator this will be "ssh".

> - **PackerRunUUID**: Current build's unique id. Can be used to specify build artifacts.

Reviewer: `Can be used to specify build artifacts.` I think it isn't clear what that means, do you think it would be a good idea to add an example here ?

> - **PackerHTTPIP**, **PackerHTTPPort**, and **PackerHTTPAddr**: HTTP IP, port, and address of the file server Packer creates to serve items in the "http" dir to the vm. The HTTP address is displayed in the format `IP:PORT`.

Reviewer: (that's probably for the future) It would be neat if those were:
```suggestion
- **packer.http.ip**, **packer.http.port**, and **packer.http.addr**: HTTP IP, port, and address of the file server Packer creates to serve items in the "http" dir to the vm. The HTTP address is displayed in the format `IP:PORT`.
```

Reviewer 2: Correct me if I'm wrong but in order to achieve this packer and http would have to be their own types, right?

Reviewer: em it would be a specific go-cty object:

```go
map[string]cty.Value{
  "packer": cty.ObjectVal(map[string]cty.Value{
    "http": cty.ObjectVal(map[string]cty.Value{
      "ip": cty.StringVal("ipvalue),
      // etc.
    },
  },
}
```

Reviewer 2: “Object” that’s the word. I was leaning towards block and end on type 😃. Thanks for the example super helpful.

- **SSHPublicKey** and **SSHPrivateKey**: The public and private key that Packer uses to connect to the instance.
  These are unique to the SSH communicator and are unset when using other communicators.

>   **SSHPublicKey** and **SSHPrivateKey** can have escape sequences and special characters so their output should be single quoted to avoid surprises. For example:

Reviewer: What do you think about making them snake case ?

Reviewer: The main argument would be that it keeps sort of a 'packer feeling' throughout the code and a simple map iteration should solve this.

Author: Like `SshPublicKey`,  `PackerHttpPort`, `Id`, etc..? 
This is going to make them different from JSON. Is that ok?

Reviewer: I think it is, a regexp should be able to fix this somewhat easily

Reviewer 2: > The main argument would be that it keeps sort of a 'packer feeling' throughout the code and a simple map iteration should solve this.

hmm, this sounds like an internal developer optimization that may result in us repeating to users in issues that variable names changed in HCL. I don't see the immediate benefit and think it should probably be held for another PR.

Reviewer: Gotcha yes I agree 🙂

  ```hcl
    provisioner "shell" {
        inline = ["echo '${build.SSHPrivateKey}' > /tmp/packer-session.pem"]
    }
  ```

> For backwards compatability, `WinRMPassword` is also available through this

Reviewer: ```suggestion
For backwards compatibility, `WinRMPassword` is also available through this
```
I always write compatability too lol

Author: I just copied and pasted it from JSON so someone else also has the same problem! But if had written I would have done the same, because I just can't spell it right too 🤦

engine, though it is no different than using the more general `Password`.

All build variables are valid to use with any of the [HCL2 functions](/docs/from-1.5/functions).
Example of using [upper](/docs/from-1.5/functions/string/upper) to upper case the build ID:

```hcl
  post-processor "shell-local" {
      inline = ["echo ${upper(build.ID)}"]
  }
```

> For builder-specific builder variables, please also refer to the builder docs.

Reviewer: Do you think it would be a good idea to add a link to the parts in the builders where this is defined ?
For example:
```suggestion
For builder-specific builder variables, please also refer to the builder docs:
* [amazon-chroot](link#part)
* ...
```
